<launch>
    <arg name="rand" default="10.0" />
    <arg name="viz" default="false" />
    <arg name="output" default="screen" />
    <arg name="cgraph_path" default="/home/xuhao/graph.dot" />
    <arg name="cgraph" default="true" />
    <arg name="enable_distance" default="true" />
    <arg name="enable_detection" default="true" />
    <arg name="enable_loop" default="true" />
    <arg name="enable_da" default="true" />
    <arg name="bag_replay" default="false" />
    <arg name="pcm_enable_debug_file" default="true" />
    <arg name="pcm_enable" default="true" />
    <arg name="debug_loop_initial_only" default="false" />
    <arg name="max_keyframe_num" default="100" />
    <arg name="debug_no_relocalization" default="false" />
    <arg name="enable_random_keyframe_deletetion" default="false" />
    <arg name="debug_no_rejection" default="false" />
    <arg name="distance_measurement_outlier_threshold" default="0.3" />
    <arg name="self_id" default="-1" />

    <node pkg="swarm_localization" name="swarm_localization" type="swarm_localization_node" output="$(arg output)"  >
        <param name="self_id" value="$(arg self_id)" type="int" />
        <param name="enable_distance" value="$(arg enable_distance)" type="bool" />
        <param name="enable_detection" value="$(arg enable_detection)" type="bool" />
        <param name="enable_loop" value="$(arg enable_loop)" type="bool" />
        <param name="enable_data_association" value="$(arg enable_da)" type="bool" />
        <param name="debug_loop_initial_only" value="$(arg debug_loop_initial_only)" type="bool" />
        <param name="debug_no_rejection" value="$(arg debug_no_rejection)" type="bool" />
        <param name="max_keyframe_num" value="$(arg max_keyframe_num)" type="int" />
        <param name="debug_no_relocalization" value="$(arg debug_no_relocalization)" type="bool" />
        <param name="enable_random_keyframe_deletetion" value="$(arg enable_random_keyframe_deletetion)" type="bool" />
        <param name="distance_measurement_outlier_threshold" value="$(arg distance_measurement_outlier_threshold)" type="double" />
        <rosparam>
            #Performance Parameters
            force_freq: 10
            thread_num: 1
            max_solver_time: 0.5

            max_accept_cost: 100
            min_keyframe_num: 1
            min_kf_movement : 0.5
            kf_use_all_nodes: false

            #Initial Configuration
            init_xy_movement : 1.5 #Debug
            init_z_movement : 0.8 #Debug 

            #Noise Models.
            vo_cov_pos_per_meter: 0.0009 #OK VO on PC, 600 width
            vo_cov_yaw_per_meter: 0.00005 #OK VO on PC, 600 width
            
            #vo_cov_pos_per_meter: 0.002 #Bad VO on PC, 600 width
            #vo_cov_yaw_per_meter: 0.00005 #Bad VO on PC, 600 width

            distance_measurement_cov: 0.00122
            

            #OUTLIER REJECTION
            det_dpos_thres: 1.0
            DA_accept_thres: 3.345
            pcm_redundant: false
            
            distance_measurement_outlier_elevation_threshold: 0.6

            loop_outlier_distance_threshold: 2.0
            #pcm_thres: 0.872 1.635 2.204 3.455 5.348 7.840
            pcm_thres: 2.8

            #Debug old thres. Not applicable to PCM.
            loop_outlier_threshold_pos: 0.5
            loop_outlier_threshold_yaw: 0.5

            #Output
            pub_swarm_odom: true
            is_pc_replay: true
            publish_full_path: false
            debug_publish_goodloops: true

            cg: #In IMU Frame.
                x: -0.06
                y: 0.0
                z: 0.0
        </rosparam>
        <param name="swarm_nodes_config" value="$(find swarm_localization)/config/swarm_nodes5.yaml" type="string" />
        <param name="initial_random_noise" value="$(arg rand)" type="double" />
        <param name="pcm_enable_debug_file" value="$(arg pcm_enable_debug_file)" type="bool" />
        <param name="pcm_enable" value="$(arg pcm_enable)" type="bool" />
        <param name="cgraph_path" value="$(arg cgraph_path)" type="string" />
        <param name="enable_cgraph_generation" value="$(arg cgraph)" type="bool" />
        <remap from="/swarm_drones/swarm_drone_fused" to="/swarm_drones/swarm_drone_fused_pc" />
        <remap from="/swarm_drones/swarm_drone_fused_relative" to="/swarm_drones/swarm_drone_fused_relative_pc" />
    </node>
    <group if="$(arg viz)">
        <node name="rvizvisualisation" pkg="rviz" type="rviz" output="log" args="-d $(find swarm_localization)/config/5drone.rviz" />
    </group>
 </launch>
